agent/reconnectingpty/reconnectingpty.go

```diff
@@ -0,0 +1,298 @@
+package reconnectingpty
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net"
+	"os/exec"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/pty"
+)
+
+const attachTimeout = 30 * time.Second
+
+// Options allows configuring the reconnecting pty.
+type Options struct {
+	// Timeout describes how long to keep the pty alive without any connections.
+	// Once elapsed the pty will be killed.
+	Timeout time.Duration
+	// Metrics tracks various error counters.
+	Metrics *prometheus.CounterVec
+	// BackendType indicates which backend to use for reconnections.
+	BackendType codersdk.ReconnectingPTYBackendType
+}
+
+// State represents the current state of the reconnecting pty.  States are
+// sequential and will only move forward.
+type State int
+
+const (
+	// StateStarting is the default/start state.  Attaching will block until the
+	// reconnecting pty becomes ready.
+	StateStarting = iota
+	// StateReady means the reconnecting pty is ready to be attached.
+	StateReady
+	// StateClosing means the reconnecting pty has begun closing.  The underlying
+	// process may still be exiting.  Attaching will result in an error.
+	StateClosing
+	// StateDone means the reconnecting pty has completely shut down and the
+	// process has exited.  Attaching will result in an error.
+	StateDone
+)
+
+type backend interface {
+	start(ctx context.Context, logger slog.Logger) error
+	attach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) (pty.PTYCmd, error)
+	close(ctx context.Context, logger slog.Logger) error
```

Comment:
the use of `error` on `close()` seems fairly arbitrary in this code.  The buffered backend returns errors closing connections but not errors closing the pty or killing the process.

Considering that the only caller of `close()` just logs the error anyway, I wonder whether it would be simpler to not return an error and have the backend be responsible for logging any problems.

Reply:
Ahh good point, I will remove the return.

I think also I was too lazy here; the process might already have exited so I did not want to return that error.  I will properly fix that along with the `cleanup` fix you mentioned above.